logpyle/__init__.py

```diff
@@ -1123,8 +1131,9 @@ def compute_watch_str(watch):
                         compute_watch_str(watch) for watch in self.watches),
                       flush=True)
 
-        ticks_per_sec = self.tick_count/max(1, time()-self.start_time)
-        self.next_watch_tick = self.tick_count + int(max(1, ticks_per_sec))
+        ticks_per_interval = (self.tick_count/max(1, time()-self.start_time)
+                         * self.watch_interval)
+        self.next_watch_tick = self.tick_count + int(max(1, ticks_per_interval))
```

Comment:
This doesn't work right if you dynamically change the watch interval. If you look at the formula, it assumes that the interval has stayed the same from the beginning.

Reply:
Hmm, I don't see it. This seems to work as intended.

Reply:
The problem is that `tick_count` is a global counter, that got built up at a rate governed by the old `watch_interval`. If you (say) double the `watch_interval` after some time of running, you'll not see another tick for about the run time that's already elapsed.

Reply:
But isn't `self.tick_count` incremented by 1 on every `tick_after` call, independent of any watches?

https://github.com/illinois-ceesd/logpyle/blob/5e0b2fcca34195f18d117dbc3dcae1dd132c681f/logpyle/__init__.py#L805-L820

Reply:
To check the units on step 1:
```
ticks/s * (s/interval) -> ticks/interval
```
That seems to make sense.

I agree with your statement of the algorithm. The problem is that if `watch_interval` changes halfway through, then the rate estimate in step 1 will be way off.

Reply:
> The problem is that if `watch_interval` changes halfway through, then the rate estimate in step 1 will be way off.

Right, but isn't the only issue with that after calling `set_watch_interval`,  the new interval only takes effect after the next time watches are printed? I don't think that is a big issue.

Reply:
I've done a small demo in examples/log.py (https://github.com/illinois-ceesd/logpyle/pull/66/commits/7d1925779e5296d033be875088b80fcbb9fd6b7f):

```python
logmgr.set_watch_interval(2) # Steps 0
logmgr.set_watch_interval(50) # Step 50
logmgr.set_watch_interval(50) # Step 1000
```

The output makes sense I think:

```
$ python examples/log.py
step.max=0 | t_sim.max=0s | t_step.max=0.0676832s | fifteen=15 | t_vis=0.055011s |
examples/log.py:57: UserWarning: Oof. Something went awry.
  warn("Oof. Something went awry.")
step.max=2 | t_sim.max=0.088648s | t_step.max=0.0271597s | fifteen=15 | t_vis=0s |
step.max=8 | t_sim.max=0.302532s | t_step.max=0.0824878s | fifteen=15 | t_vis=0s |
step.max=26 | t_sim.max=1.41205s | t_step.max=0.07392s | fifteen=15 | t_vis=0s |
step.max=54 | t_sim.max=2.75923s | t_step.max=0.08833s | fifteen=15 | t_vis=0s |
step.max=838 | t_sim.max=45.1587s | t_step.max=0.0913711s | fifteen=15 | t_vis=0s |
step.max=1621 | t_sim.max=88.2842s | t_step.max=0.0891988s | fifteen=15 | t_vis=0s |
step.max=1698 | t_sim.max=92.5796s | t_step.max=0.0233502s | fifteen=15 | t_vis=0s |
step.max=1775 | t_sim.max=96.3921s | t_step.max=0.0772991s | fifteen=15 | t_vis=0s |
step.max=1853 | t_sim.max=100.442s | t_step.max=0.0352693s | fifteen=15 | t_vis=0s |
step.max=1931 | t_sim.max=104.614s | t_step.max=0.0506461s | fifteen=15 | t_vis=0s |
```

Reply:
This is great, btw. Thanks!

Reply:
> ie, the changed interval is only effective after step 1621 (while the function is called at step 1000), but works as expected after step 1621.

Right... that's the bug I'm pointing out!

Reply:
What do you think?

```
$ python examples/log.py
step.max=0 | t_sim.max=0s | t_step.max=0.135209s | fifteen=15 | t_vis=0.0550127s |
step.max=2 | t_sim.max=0.142541s | t_step.max=0.055336s | fifteen=15 | t_vis=0s |
step.max=8 | t_sim.max=0.427875s | t_step.max=0.0244982s | fifteen=15 | t_vis=0s |
examples/log.py:57: UserWarning: Oof. Something went awry.
  warn("Oof. Something went awry.")
step.max=26 | t_sim.max=1.18263s | t_step.max=0.0629508s | fifteen=15 | t_vis=0s |
step.max=843 | t_sim.max=46.5442s | t_step.max=0.0823319s | fifteen=15 | t_vis=0s |
step.max=1000 | t_sim.max=54.7779s | t_step.max=0.123746s | fifteen=15 | t_vis=0.0550847s |
step.max=1077 | t_sim.max=58.6151s | t_step.max=0.0705163s | fifteen=15 | t_vis=0s |
step.max=1154 | t_sim.max=63.135s | t_step.max=0.0817201s | fifteen=15 | t_vis=0s |
step.max=1231 | t_sim.max=67.8484s | t_step.max=0.019572s | fifteen=15 | t_vis=0s |
step.max=1308 | t_sim.max=72.1498s | t_step.max=0.0242062s | fifteen=15 | t_vis=0s |
step.max=1385 | t_sim.max=76.793s | t_step.max=0.0640011s | fifteen=15 | t_vis=0s |
step.max=1461 | t_sim.max=81.3177s | t_step.max=0.027359s | fifteen=15 | t_vis=0s |
step.max=1537 | t_sim.max=85.492s | t_step.max=0.0293491s | fifteen=15 | t_vis=0s |
step.max=1613 | t_sim.max=89.6905s | t_step.max=0.0713189s | fifteen=15 | t_vis=0s |
step.max=1689 | t_sim.max=93.6382s | t_step.max=0.0671051s | fifteen=15 | t_vis=0s |
step.max=1765 | t_sim.max=97.8387s | t_step.max=0.0485611s | fifteen=15 | t_vis=0s |
step.max=1841 | t_sim.max=102.326s | t_step.max=0.0839317s | fifteen=15 | t_vis=0s |
step.max=1917 | t_sim.max=106.608s | t_step.max=0.0868111s | fifteen=15 | t_vis=0s |
step.max=1993 | t_sim.max=110.716s | t_step.max=0.0415459s | fifteen=15 | t_vis=0s |
```

Differences:
- No watch print at step ~50 since set_watch_interval set a new timer
- Watches start printing at step 1000 with a 5 second interval

Reply:
This works for me.  A fixed interval would be cool, but allowing it to change is icing. I don't have the same concerns about the steadiness of the output that you guys don't seem to like.